rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+                          newLeaf,
+                          existingTail,
+                          existingHead,
+                          existingPointer,
+                          prefixPath
+                        )
+        } yield dividedSkip._1
+
+      case TriePath(elems :+ Node(ptr), Some(Skip(affix, existingPointer)), common) => // split existing skip
+        val incomingPath                 = currentPath.drop(common.size)
+        val affixBytes                   = affix.toSeq.toList
+        val prefixPath                   = commonPrefix(incomingPath, affixBytes)
+        val incomingHead :: incomingTail = incomingPath.drop(prefixPath.size)
+        val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+
+        for {
+          dividedSkip <- divideSkip(
+                          incomingTail,
+                          incomingHead,
+                          newLeaf,
+                          existingTail,
+                          existingHead,
+                          existingPointer,
+                          prefixPath
+                        )
+          (newElements, topTrie) = dividedSkip
+          pointerBlock           <- pointerBlockStore.get(ptr)
+          updatedPointerBlock    = pointerBlock.get.updated((toInt(common.last), topTrie) :: Nil)
+          nodeToExistingPB       <- storePointerBlock(updatedPointerBlock)
+          result                 <- rehash(common.init, elems, nodeToExistingPB, newElements)
+        } yield result
+
+      case TriePath(elems :+ Node(ptr), None, common) => // add to existing node
+        for {
+          pastPB <- pointerBlockStore.get(ptr).map(_.get)
+          key    = currentPath.drop(common.size)
+          s      = skipOrValue(key, newLeaf)
+          newPB  = pastPB.updated((toInt(common.last), s) :: Nil)
+          node   <- storePointerBlock(newPB)
+          result <- rehash(common.init, elems, node, newLeaf :: s :: Nil)
+        } yield result
+
+      case TriePath(elems :+ (_: Leaf), _, path) => // update value
+        rehash(path, elems, newLeaf, Nil)
+      case _ => Sync[F].raiseError(new RuntimeException("malformed trie"))
+    }
+
+  private def rehash(
+      path: KeyPath,
+      rest: Vector[Trie],
+      lastSeen: Trie,
+      acc: List[Trie]
+  ): F[List[Trie]] = {
+    type Params = (KeyPath, Vector[Trie], Trie, List[Trie])
+    def go(params: Params): F[Either[Params, List[Trie]]] = params match {
+      case (
+          currentPath: KeyPath,
+          currentRest: Vector[Trie],
+          nextLastSeen: Trie,
+          currentAcc: List[Trie]
+          ) =>
+        currentRest match {
+          case head :+ Node(ptr) =>
+            for {
+              pointerBlock <- pointerBlockStore
+                               .get(ptr)
+                               .map(_.get.updated((toInt(currentPath.last), nextLastSeen) :: Nil))
+              nodeToPB <- storePointerBlock(pointerBlock)
```

Comment:
Since we return all the modified Trie-s to be saved in `process`, Trie has its recursion broken by the PointerBlocks, maybe it'd make sense to move all the `storePointerBlock` calls in there?

Oh wait... There isn't a PointerBlock in the Node, there's a PointerBlockPointer...

Maybe making the PointerBlock sparse would allow its fusion with the Trie.Node. But that's just an IDEA for future.

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3227